src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorFileChangeDetector.cs

```diff
@@ -130,11 +145,69 @@ protected virtual IReadOnlyList<string> GetExistingRazorFiles(string workspaceDi
             return existingRazorFiles.ToArray();
         }
 
-        private void FileSystemWatcher_RazorFileEvent_Background(string physicalFilePath, RazorFileChangeKind kind)
+        // Internal for testing
+        internal void FileSystemWatcher_RazorFileEvent_Background(string physicalFilePath, RazorFileChangeKind kind)
         {
-            Task.Factory.StartNew(
-                () => FileSystemWatcher_RazorFileEvent(physicalFilePath, kind),
-                CancellationToken.None, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+            lock (_pendingNotificationsLock)
+            {
+                if (!_pendingNotifications.TryGetValue(physicalFilePath, out var currentNotification))
+                {
+                    currentNotification = new DelayedFileChangeNotification();
+                    _pendingNotifications[physicalFilePath] = currentNotification;
+                }
+
+                if (currentNotification.ChangeKind != null)
+                {
+                    Debug.Assert(currentNotification.ChangeKind == RazorFileChangeKind.Added || currentNotification.ChangeKind == RazorFileChangeKind.Removed);
+
+                    if (currentNotification.ChangeKind != kind)
+                    {
+                        // Previous was added and current is removed OR previous was removed and current is added. Either way there's no
+                        // actual change to notify, null it out.
+                        currentNotification.ChangeKind = null;
```

Comment:
How likely is it that we'll get a Removed/Added notification in quick succession but have different content than originally was? (Maybe the document changing in another IDE). Do we have to worry about those?

Reply:
Interesting point. It might be something we have to worry about but I'm making the assumption that we'll be notified about it via other channels like our textDocument/didChange notification handling